extension-common/pxfconfigparser.c

```diff
@@ -0,0 +1,661 @@
+ * Expect the provided cluster to be a valid pointer to pxf_cluster struct
+ * Expect group_name to be the second token in key
+ * Expect property_name to be the third token in key
+ * Expect ctx to be a fully-populated parser_line_context
+ */
+static void
+populate_group_property(pxf_cluster *cluster, char *group_name, char *property_name, const parser_line_context *ctx)
+{
+    char *value_copy = pstrdup(ctx->value);
+    // Find the group by name
+    pxf_service_group *group = get_pxf_service_group_in_cluster(cluster, group_name);
+    if (group == NULL)
+    {
+        group = palloc0(sizeof(pxf_service_group));
+        group->name = pstrdup(group_name);
+        cluster->groups = lappend(cluster->groups, group);
+    }
+    if (strcmp(property_name, "ports") == 0)
+    {
+        List *ports = get_tokens(value_copy, ",");
+        if (ports == NULL)
+        {
+            // Some ports are empty strings
+            ereport(ERROR, (errcode(ERRCODE_CONFIG_FILE_ERROR),
+                            errmsg("configuration property '%s' contains invalid value '%s', "
+                                   "at line %d of file '%s'",
+                                   ctx->key, ctx->value, ctx->line_number, ctx->filepath),
+                            errhint("please make sure the value is valid port numbers separated by ','")));
+        }
+
+        ListCell *lc;
+        foreach (lc, ports)
+        {
+            int port = parse_int(lfirst(lc), ctx);
+            group->ports = lappend_int(group->ports, port);
+        }
+    }
+    else if (strcmp(property_name, "ssl") == 0)
+    {
+        group->is_ssl = parse_bool(value_copy, ctx);
+    }
+    else
+    {
+        ereport(DEBUG1,
+                (errmsg("ignoring unknown configuration property '%s' in config file '%s' at line %d", ctx->key, ctx->filepath, ctx->line_number)));
+    }
+
+    pfree(value_copy);
+}
+
+/*
+ * Add a new pxf_cluster entry with a name of cluster_name into the deployment
+ * Expect deployment to be a valid pointer to pxf_deployment
+ * Expect cluster_name to be the same as the cluster's folder name
+ */
+static pxf_cluster *
+create_pxf_cluster(pxf_deployment *deployment, const char *cluster_name)
+{
+    pxf_cluster *cluster = palloc0(sizeof(pxf_cluster));
+    cluster->name = pstrdup(cluster_name);
+
+    deployment->clusters = lappend(deployment->clusters, cluster);
+    return cluster;
+}
+
+/*
+ * Clean up all allocated memory in pxf_deployment struct instance
+ * This function should be called when the pxf_deployment struct is no longer needed
+ */
+void clean_pxf_deployment(pxf_deployment *deployment)
+{
+    if (deployment == NULL)
+    {
+        return;
+    }
+    if (deployment->clusters != NULL)
+    {
+        ListCell *lc;
+        foreach (lc, deployment->clusters)
+        {
+            pxf_cluster *cluster = lfirst(lc);
+            clean_pxf_cluster(cluster);
+            pfree(cluster);
+        }
+        list_free(deployment->clusters);
```

Comment:
```suggestion
        }
        list_free_deep(deployment->clusters);
```
I think you can do a deep clean of the list to clean cluster pointers